### PR TITLE
Installation on the fly of DALI whl when building the TF plugin

### PR DESCRIPTION
**Pull request by jantonguirao**
co-author: @DEKHTIARJonathan 

#### Why we need this PR?
*Pick one, remove the rest*
- It allows DALI TF to resolve dependencies during package build, allowing the user to install DALI TF in a single pip install call

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Added a custom egg_info step that checks whether DALI is already installed and installs it using pip if not.*
     *After reinstall, the program needs to be relaunched to be able to import the package*
 - Affected modules and functionalities:
     *DALI TF setup.py.in*
 - Key points relevant for the review:
     *Custom dependency installation code*
 - Validation and testing:
     *TODO ?*
 - Documentation (including examples):
     *TODO ?*

**JIRA TASK**: *[DALI-1559]*

### CODE REVIEW

File: dali_tf_plugin/setup.py.in [hunk 1]
```diff
@@ -14,16 +14,64 @@
...
+
+DALI_PKG_REQ = 'nvidia-dali@DALI_FLAVOR_MINUS@-cuda@CUDA_VERSION_SHORT_DIGIT_ONLY@==@DALI_VERSION@'
+NVIDIA_INDEX_URL = "https://developer.download.nvidia.com/compute/redist/@DALI_FLAVOR@"
+
+def is_dali_installed():
+    try:
+        import nvidia.dali
+        return True
+    except:
+        return False
```
Comment (JanuszL) [inline dali_tf_plugin/setup.py.in:33]:
```suggestion
    try:
        import nvidia.dali
    except:
        return False
     return True
```

**Comment by jantonguirao** [comment]
!build

File: dali_tf_plugin/setup.py.in [hunk 2]
```diff
@@ -14,16 +14,64 @@
...
+def pip_install(pkg_req, user=False):
+    pip_args = [
+        "install",
+        "--index-url", NVIDIA_INDEX_URL
+    ]
+    if user:
+        pip_args.append("--user")
+    pip_args.append(pkg_req)
+
+    from pip._internal import main as pip
```
Comment (JanuszL) [inline dali_tf_plugin/setup.py.in:44]: `__internal ` doesn't sound like a public PIP api. https://github.com/NVIDIA/DALI/blob/master/qa/setup_packages.py#L7L13 shows that `_internal` can be changed version to version

File: dali_tf_plugin/setup.py.in [hunk 3]
```diff
@@ -14,16 +14,64 @@
...
+NVIDIA_INDEX_URL = "https://developer.download.nvidia.com/compute/redist/@DALI_FLAVOR@"
+
+def is_dali_installed():
+    try:
+        import nvidia.dali
+        return True
+    except:
+        return False
+
+def pip_install(pkg_req, user=False):
```
Comment (JanuszL) [inline dali_tf_plugin/setup.py.in:35]:
```suggestion
def pip_install_dali(pkg_req, user=False):
```

File: dali_tf_plugin/setup.py.in (hunk 2 above)
Comment (jantonguirao) [reply dali_tf_plugin/setup.py.in:44]: Yes, I was worried about that too. Another option is to invoke pip program directly

File: dali_tf_plugin/setup.py.in [hunk 4]
```diff
@@ -14,16 +14,64 @@
...
+        """Installing DALI if not already installed during EggInfo command (first egg_info run)"""
+        handle_deps = all([keyword in sys.argv for keyword in ["egg_info", "--egg-base"]]) 
+        if handle_deps and not is_dali_installed():
+            ret = pip_install(DALI_PKG_REQ, user=False)
+            if not ret:
+                ret = pip_install(DALI_PKG_REQ, user=True)
+            if not ret:
+                raise RuntimeError("Could not install {}".format(DALI_PKG_REQ))
+            # For nvidia.dali to be importable, restart the program from scratch
+            os.execl(sys.executable, sys.executable, *sys.argv)
```
Comment (JanuszL) [inline dali_tf_plugin/setup.py.in:71]: Maybe it is enough to reaload `setuptools` or other modules (see https://www.geeksforgeeks.org/reloading-modules-python)?

Comment (jantonguirao) [reply dali_tf_plugin/setup.py.in:71]: reload is to reload a module that was already imported. In this case, when I try to import `nvidia.dali` I get an error saying that there's no such package. When you rerun the program the package is available. I am assuming that when we start python the available packages are loaded into memory. Not sure if we can trigger reloading that, so far I haven't found any info about it.

**Comment by jantonguirao** [comment]
!build

**Comment by klecki** [comment]
Hmm, I wonder why this doesn't work out of the box: https://pip.pypa.io/en/stable/reference/pip_install/#installation-order. 

Also, there are other options like --root or --prefix: 
https://pip.pypa.io/en/stable/reference/pip_install/#cmdoption-root
https://pip.pypa.io/en/stable/reference/pip_install/#cmdoption-prefix
that will probably break if we decide to install DALI on our own (unless we uninstall it or install to a temporary directory, compile the plugin and remove).

Also, what with specific versions? If users wants some older DALI, we need to pass it to the install. If the user makes a mistake and specifies different versions, depending on the combination it will probably work in *some* cases.

Also, doesn't the same problem appear when the user tries to install TensorFlow and DALI tf plugin?

**Comment by DEKHTIARJonathan** [comment]
@klecki 
> Hmm, I wonder why this doesn't work out of the box: https://pip.pypa.io/en/stable/reference/pip_install/#installation-order.

it's actually very simple, I already explained it on Slack. Here are in order the different commands executed:

```
egg_info
bdist_wheel
build
build_py
egg_info (again yes)
build_ext  # Where you compile your extension
install        # After the compilation, so installing dependencies here won't work
install_lib
install_egg_info
install_scripts
```

You can see that the 2nd operation build the wheel and therefore also compiles the tensorflow extension.
Dependencies are only installed during the "install" phase. Way later in the process.

So the only solution is to install `nvidia-dali` during the first `egg-info`.

**Comment by jantonguirao** [comment]
Closing as it is no longer necessary to have nvidia-dali installed at the time when the TF plugin is built

**Pull request closed without merging**